### ISSUE TITLE
[WebAssembly] Support `tailcc` calling convention in wasm backend

### ISSUE THREAD

**Issue by 8051Enthusiast** [llvm/llvm-project#90403]
Currently, llvm's wasm backend does not support the `tailcc` calling convention even when tail calls are supported.
I thought about making a PR that just adds it to the `callingConvSupported` function, but then I wasn't sure if it should handle the absense of the tail-call feature or fast-isel not emitting tail calls.

Example program:
```
target triple = "wasm32-unknown-unknown"
declare tailcc void @void_nullary()

define void @tailcc_tail_call_void_nullary() {
  tail call tailcc void @void_nullary()
  ret void
}
```
`llc` output:
```
error: <unknown>:0:0: in function tailcc_tail_call_void_nullary void (): WebAssembly doesn't support language-specific or target-specific calling conventions yet
```

**Comment by dschuff** [comment]
According to the langref, "tailcc" has guaranteed TCO... well actually, it says "guarantee that tail calls will be produced whenever possible". So depending on whether "possible" is from the perspective of the frontend/producer or the backend, maybe it's not so much of a guarantee at all 😅.
In any case, in order to make this work, we'd need to get a better handle on exactly what the guarantees and expectations are, and how compatible those are with the capabilities of wasm's tail call instruction. (having some involvement from someone who works on or uses a language that uses this feature would be really helpful here, so we know whether it's working... maybe that's you?). 
Regarding the more-specific questions, probably we'd just have to continue raising an error if this calling convention is used without the tail call feature being enabled (since the backend isn't able to make it actually work). We wouldn't necessarily need to add fast-isel support, because there's always the ability to bail out of fast isel and select a block with selection-dag (although it would still be good to support because if practically every block needs to bail out, then compilation gets fairly slow even in unoptimized/debug mode).

**Comment by 8051Enthusiast** [comment]
I looked a bit more into how `tailcc` is supposed to work:
> According to the langref, "tailcc" has guaranteed TCO... well actually, it says "guarantee that tail calls will be produced whenever possible". So depending on whether "possible" is from the perspective of the frontend/producer or the backend, maybe it's not so much of a guarantee at all 😅.

Yeah, in my opinion that guarantee is not that useful. It is still be possible to call tailcc functions when the conditions are not met, so the I think the only effect of that guarantee is that the `tail` attribute is not just taken as a hint but should always be tail-called if possible, even if it has worse performance. But if someone wants to have guaranteed tail calls, they'd use `musttail`.

> In any case, in order to make this work, we'd need to get a better handle on exactly what the guarantees and expectations are, and how compatible those are with the capabilities of wasm's tail call instruction. (having some involvement from someone who works on or uses a language that uses this feature would be really helpful here, so we know whether it's working... maybe that's you?).

I think the most important point of `tailcc` is this part from the langref, especially the second point:
> In addition, if the calling convention is not swifttailcc or tailcc:
>  * All ABI-impacting function attributes, such as sret, byval, inreg, returned, and inalloca, must match.
> * The caller and callee prototypes must match. Pointer types of parameters or return types may differ in pointee type, but not in address space.

Having the function signatures of the callee and caller match is a big constraint and I think the main point of `tailcc` is to have different conventions about eg who pops the arguments so that tail calls across different signatures are possible.
I'm actually not that well-versed in the specifics of the wasm ABI, but if I understood it correctly, the `return_call` should pop the arguments of the callee and set up the arguments of the caller so that a mismatch in function signature should not matter. Again I don't know the specifics, so correct me if I'm wrong.

> We wouldn't necessarily need to add fast-isel support, because there's always the ability to bail out of fast isel and select a block with selection-dag (although it would still be good to support because if practically every block needs to bail out, then compilation gets fairly slow even in unoptimized/debug mode).

I looked at how another backend did it (I think x86?) and it just aborted fast isel if it encountered tailcc so that's probably what the wasm backend should do as well.

**Comment by dschuff** [comment]
> Having the function signatures of the callee and caller match is a big constraint and I think the main point of `tailcc` is to have different conventions about eg who pops the arguments so that tail calls across different signatures are possible. I'm actually not that well-versed in the specifics of the wasm ABI, but if I understood it correctly, the `return_call` should pop the arguments of the callee and set up the arguments of the caller so that a mismatch in function signature should not matter. Again I don't know the specifics, so correct me if I'm wrong.

Yes, this is right. And actually I should have mentioned this up front, but the wasm backend does support `musttail` if you have the tail call feature enabled. So that might be enough for your purposes even without `tailcc`

**Comment by 8051Enthusiast** [comment]
Oh yeah, I actually added `musttail` to my compiler just yesterday (in previous llvm versions, there were no C bindings for setting `musttail`), but with this the IR verifier complains about mismatched function signatures of the caller and callee if the calling convention is not `tailcc`, which is rather unfortunate.
